File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/ResponseString.java
package spectable.common;

import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class ResponseString {
    private static final String DNCString = "?DNC?";

    public ResultString result;

    public ResponseString(ResultString result) {
        this.result = result;
    }

    /** Builds from the text form, e.g. Money as "25 USD". */
    public static ResponseString fromText(String text) {
        java.util.List<String> parts = Tokens.require(text, 1, "Response");
        return new ResponseString(ResultString.fromText(parts.get(0)));
    }

    public ResponseString(String text) {
        ResponseString parsed = fromText(text);
        this.result = parsed.result;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ResponseString)) return false;
        ResponseString that = (ResponseString) o;
        return Objects.equals(result, that.result);
    }

    @Override
    public int hashCode() {
        return Objects.hash(result);
    }

    @Override
    public String toString() {
        return Tokens.nested(String.valueOf(result));
    }
}

File: JavaScript/ExampleTests/common/ResponseString.js
import { ResultString } from "./ResultString.js";
import * as tokens from "./tokens.js";

export class ResponseString {
  static DNC_STRING = "?DNC?";

  constructor(result = "") {
    this.result = result;
  }

  static fromList(values) {
    const v = Array.from(values);
    return new ResponseString(
      v[0] ?? ""
    );
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text) {
    const parts = tokens.require_(text, 1, "Response");
    return new ResponseString(
      ResultString.fromText(parts[0])
    );
  }

  toString() {
    return tokens.nested(String(this.result));
  }

  equals(other) {
    if (!(other instanceof ResponseString)) return false;
    return this.result.equals(other.result);
  }

  static equalLists(a, b) {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]));
  }
}
